src/sockaddr.rs

```diff
@@ -168,34 +199,88 @@ impl SockAddr {
     }
 }
 
-// SocketAddrV4 and SocketAddrV6 are just wrappers around sockaddr_in and sockaddr_in6
-
-// check to make sure that the sizes at least match up
-fn _size_checks(v4: SocketAddrV4, v6: SocketAddrV6) {
-    unsafe {
-        mem::transmute::<SocketAddrV4, sockaddr_in>(v4);
-        mem::transmute::<SocketAddrV6, sockaddr_in6>(v6);
-    }
-}
-
 impl From<SocketAddrV4> for SockAddr {
     fn from(addr: SocketAddrV4) -> SockAddr {
-        unsafe {
-            SockAddr::from_raw_parts(
-                &addr as *const _ as *const _,
-                mem::size_of::<SocketAddrV4>() as socklen_t,
-            )
+        #[cfg(unix)]
+        let sin_addr = in_addr {
+            s_addr: u32::from_ne_bytes(addr.ip().octets()),
+        };
+        #[cfg(windows)]
+        let sin_addr = unsafe {
+            let mut s_un = mem::zeroed::<in_addr_S_un>();
+            *s_un.S_addr_mut() = u32::from_ne_bytes(addr.ip().octets());
+            in_addr { S_un: s_un }
+        };
+
+        let mut storage = MaybeUninit::<sockaddr_storage>::uninit();
+        let sockaddr_in = unsafe { &mut *(storage.as_mut_ptr() as *mut sockaddr_in) };
```

Comment:
I realized it's not very clean or safe to take out a safe Rust reference to the uninitialized memory region. A much better way is of course to use `storage.as_mut_ptr().write(value)` to guarantee that we just write a value into the region and don't use the reference for anything else that might be UB

Reply:
Yes, the old code had UB. I had a comment about it but it's gone. The solution using `write` is perfect.